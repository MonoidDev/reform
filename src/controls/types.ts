import { Either } from '../types/Either';
import { ErrorMessage } from '../types/ErrorMessage';
import type { FormControl } from './FormControl'
import type { StructFormControl } from './StructFormControl'

export type AnyFormControl = FormControl<any, any, any, any> | StructFormControl<any>;

export type FormResult<O> = Either<ErrorMessage | undefined, O>;

export type Unsubscribe = () => void;
