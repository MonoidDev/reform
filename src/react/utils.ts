import { useState } from 'react';

export const useForceUpdate = () => {
  const [, s] = useState({});
  return () => s({});
}
