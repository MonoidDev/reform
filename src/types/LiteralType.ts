import { makeLeft, makeRight } from './Either';
import { ErrorMessage } from './ErrorMessage';
import { makeResolver, Resolver } from './Resolver';

export type LiteralValue = string | number | boolean | null | undefined;

export interface TLiteral<T extends LiteralValue>
  extends Resolver<unknown, T, ErrorMessage, 'literal'> {
  value: T;
}

export function literal<T extends LiteralValue>(value: T): TLiteral<T> {
  return {
    ...makeResolver<unknown, T, ErrorMessage, 'literal'>('literal', (input) => {
      if (input === value) {
        return makeRight(input as T);
      } else {
        return makeLeft({
          message: `Expect ${value}, got ${input}. `,
        });
      }
    }),
    value,
  };
}
