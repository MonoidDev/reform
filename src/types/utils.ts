export const isObject = (input: unknown): input is {} => {
  return (
    (typeof input === 'object' || typeof input === 'function') && input !== null
  );
};
