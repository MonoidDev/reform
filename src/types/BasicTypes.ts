import { isLeft, makeLeft, makeRight } from './Either';
import { ErrorMessage } from './ErrorMessage';
import { AnyResolver, OutputOf } from './infer';
import {
  ErrorContext,
  makeErrorContext,
  makeResolver,
  Resolver,
} from './Resolver';
import { isObject } from './utils';

export interface TAny extends Resolver<any, any, ErrorMessage, 'any'> {}

export function any(): TAny {
  return {
    ...makeResolver<any, any, ErrorMessage, 'any'>('any', (input) =>
      makeRight(input)
    ),
  };
}

export interface TString
  extends Resolver<unknown, string, ErrorMessage, 'string'> {}

export function string(): TString {
  return {
    ...makeResolver<unknown, string, ErrorMessage, 'string'>(
      'string',
      (input) => {
        if (typeof input === 'string') {
          return makeRight(input);
        } else {
          return makeLeft({
            message: `Expect a string, got ${typeof input}. `,
          });
        }
      }
    ),
  };
}

export interface TNumber
  extends Resolver<unknown, number, ErrorMessage, 'number'> {}

export function number(): TNumber {
  return {
    ...makeResolver<unknown, number, ErrorMessage, 'number'>(
      'number',
      (input) => {
        if (typeof input === 'number') {
          return makeRight(input);
        } else {
          return makeLeft({
            message: `Expect a number, got ${typeof input}. `,
          });
        }
      }
    ),
  };
}

export interface TBoolean
  extends Resolver<unknown, boolean, ErrorMessage, 'boolean'> {}

export function boolean(): TBoolean {
  return {
    ...makeResolver<unknown, boolean, ErrorMessage, 'boolean'>(
      'boolean',
      (input) => {
        if (typeof input === 'boolean') {
          return makeRight(input);
        } else {
          return makeLeft({
            message: `Expect a boolean, got ${typeof input}. `,
          });
        }
      }
    ),
  };
}

export interface TNull extends Resolver<unknown, null, ErrorMessage, 'null'> {}

export function nullType(): TNull {
  return {
    ...makeResolver<unknown, null, ErrorMessage, 'null'>('null', (input) => {
      if (input === null) {
        return makeRight(input);
      } else {
        return makeLeft({
          message: `Expect a null, got ${typeof input}. `,
        });
      }
    }),
  };
}

export interface TUndefined
  extends Resolver<unknown, undefined, ErrorMessage, 'undefined'> {}

export function undefinedType(): TUndefined {
  return {
    ...makeResolver<unknown, undefined, ErrorMessage, 'undefined'>(
      'undefined',
      (input) => {
        if (input === undefined) {
          return makeRight(input);
        } else {
          return makeLeft({
            message: `Expect an undefined, got ${typeof input}. `,
          });
        }
      }
    ),
  };
}

export interface TObject
  extends Resolver<unknown, {}, ErrorMessage, 'object'> {}

export function object(): TObject {
  return {
    ...makeResolver<unknown, {}, ErrorMessage, 'object'>('object', (input) => {
      if (isObject(input)) {
        return makeRight(input);
      } else {
        return makeLeft({
          message: `Expect an object, got ${typeof input}. `,
        });
      }
    }),
  };
}

export interface TArray<T extends AnyResolver>
  extends Resolver<unknown, OutputOf<T>[], ErrorMessage, 'array'> {}

export function array<T extends AnyResolver>(t: T): TArray<T> {
  return {
    ...makeResolver<unknown, OutputOf<T>[], ErrorMessage, 'array'>('array'),
    resolve(input, contexts = [makeErrorContext()]) {
      if (!Array.isArray(input)) {
        contexts[contexts.length - 1].error = {
          message: `Expect an array, got ${typeof input}`,
        };
        return makeLeft(contexts);
      }

      let value: any = [];

      const newErrorContexts: ErrorContext[] = []; 
      for (let i = 0; i < input.length; i++) {
        const nextContext: ErrorContext = {
          inputPath: [...contexts[contexts.length - 1].inputPath, i],
          resolverPath: [...contexts[contexts.length - 1].resolverPath],
          error: undefined,
        };

        const result = t.resolve(input[i], [nextContext]);

        if (isLeft(result)) {
          newErrorContexts.push(...result.left);
        } else {
          value.push(result.right);
        }
      }

      if (newErrorContexts.length) {
        contexts.pop(); // Remove last context because there're no errors
        return makeLeft([
          ...contexts,
          ...newErrorContexts,
        ]);
      }

      return makeRight(value);
    }
  };
}
