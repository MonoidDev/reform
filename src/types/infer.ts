import { TAny } from "./BasicTypes"
import { Resolver } from "./Resolver"

export type AnyResolver = Resolver<any, any, any, string>;

export type ResolverMap = {
  [K in string]: AnyResolver;
}

export type InputOfResolverMap<RM extends ResolverMap> = {
  [K in keyof RM]: InputOf<RM[K]>;
}

export type OutputOfResolverMap<RM extends ResolverMap> = {
  [K in keyof RM]: OutputOf<RM[K]>;
}

export type InputOf<T extends AnyResolver> = T extends Resolver<infer I, any, any, any> ? I : never;

export type OutputOf<T extends AnyResolver> = T extends Resolver<any, infer O, any, any> ? O : never;
