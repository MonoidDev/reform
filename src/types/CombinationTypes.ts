import { isRight, makeLeft } from './Either';
import { ErrorMessage } from './ErrorMessage';
import { AnyResolver, OutputOf } from './infer';
import {
  ErrorContext,
  makeErrorContext,
  makeResolver,
  Resolver,
} from './Resolver';

export interface TUnion<CS extends [AnyResolver, ...AnyResolver[]]>
  extends Resolver<unknown, OutputOf<CS[number]>, ErrorMessage, 'union'> {
  types: CS;
}

export function union<CS extends [AnyResolver, ...AnyResolver[]]>(
  cs: CS,
): TUnion<CS> {
  return {
    ...makeResolver<unknown, OutputOf<CS[number]>, ErrorMessage, 'union'>(
      'union',
    ),
    resolve(input, contexts = [makeErrorContext()]) {
      const newErrorContexts: ErrorContext[] = [];
      let i = 0;
      for (const resolver of cs) {
        const nextContext: ErrorContext = {
          inputPath: [...contexts[contexts.length - 1].inputPath],
          resolverPath: [...contexts[contexts.length - 1].resolverPath, i],
          error: undefined,
        };

        const result = resolver.resolve(input, [nextContext]);

        if (isRight(result)) {
          return result;
        } else {
          newErrorContexts.push(...result.left);
        }
        i++;
      }

      contexts.pop();
      return makeLeft([...contexts, ...newErrorContexts]);
    },
    types: cs,
  };
}
