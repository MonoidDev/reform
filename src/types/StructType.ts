import { isLeft, makeLeft, makeRight } from './Either';
import { ErrorMessage } from './ErrorMessage';
import { InputOfResolverMap, OutputOfResolverMap, ResolverMap } from './infer';
import {
  ErrorContext,
  makeErrorContext,
  makeResolver,
  Resolver,
} from './Resolver';
import { isObject } from './utils';

export interface TStruct<RM extends ResolverMap>
  extends Resolver<
    InputOfResolverMap<RM>,
    OutputOfResolverMap<RM>,
    ErrorMessage,
    'struct'
  > {
  props: RM;
}

export function struct<RM extends ResolverMap>(props: RM): TStruct<RM> {
  return {
    ...makeResolver('struct'),
    resolve(input, contexts = [makeErrorContext()]) {
      if (!isObject(input)) {
        contexts[contexts.length - 1].error = {
          message: `Expect an object, got ${typeof input}`,
        };
        return makeLeft(contexts);
      }

      let value: any = {};

      const newErrorContexts: ErrorContext[] = [];

      for (const [key, resolver] of Object.entries(props)) {
        const nextContext: ErrorContext = {
          inputPath: [...contexts[contexts.length - 1].inputPath, key],
          resolverPath: [...contexts[contexts.length - 1].resolverPath, key],
          error: undefined,
        };
        const result = resolver.resolve(input[key], [nextContext]);

        if (isLeft(result)) {
          newErrorContexts.push(...result.left);
        } else {
          value[key] = result.right;
        }
      }

      if (newErrorContexts.length) {
        contexts.pop(); // Remove last context because there're no errors
        return makeLeft([...contexts, ...newErrorContexts]);
      }

      return makeRight(value);
    },
    props,
  };
}
