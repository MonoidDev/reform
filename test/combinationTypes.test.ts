import { literal } from '../src/types/LiteralType';

import { union } from '../src/types/CombinationTypes';
import { OutputOf } from '../src/types/infer';
import { expectLeft, expectRight } from './utils';
import { struct } from '../src/types/StructType';
import { string } from '../src/types/BasicTypes';

describe('union', () => {
  const enumerateSimple = union([literal('A')]);

  const enumerate = union([
    literal('A'),
    literal('B'),
    literal('C'),
    literal('D'),
  ]);

  it('can type', () => {
    const v1: OutputOf<typeof enumerateSimple> = 'A';  
    // @ts-expect-error
    const v2: OutputOf<typeof enumerateSimple> = 'B';

    const v3: OutputOf<typeof enumerate> = 'A';
    const v4: OutputOf<typeof enumerate> = 'B';
    const v5: OutputOf<typeof enumerate> = 'C';
    const v6: OutputOf<typeof enumerate> = 'D';

    // @ts-expect-error
    const v7: OutputOf<typeof enumerate> = '';
  });

  it('can reflect', () => {
    expect(enumerate.types[0].value).toBe('A');
    expect(enumerate.types[1].value).toBe('B');
    expect(enumerate.types[2].value).toBe('C');
    expect(enumerate.types[3].value).toBe('D');
  });

  it('can validate', () => {
    expectRight(enumerate.resolve('A')).toBe('A');
    expectRight(enumerate.resolve('B')).toBe('B');

    expectLeft(enumerate.resolve('')).toMatchSnapshot();
    expectLeft(enumerate.resolve({})).toMatchSnapshot();

    const nested = struct({
      gender: union([
        literal('M'),
        literal('F'),
        literal('Other'),
      ]),
      name: string(),
    });

    expectLeft(nested.resolve({
      gender: '',
      name: 0,
    })).toMatchSnapshot();
  });
});
