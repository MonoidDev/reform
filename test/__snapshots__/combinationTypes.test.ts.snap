// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`union can validate 1`] = `
Array [
  Object {
    "error": Object {
      "message": "Expect A, got . ",
    },
    "inputPath": Array [],
    "resolverPath": Array [
      0,
    ],
  },
  Object {
    "error": Object {
      "message": "Expect B, got . ",
    },
    "inputPath": Array [],
    "resolverPath": Array [
      1,
    ],
  },
  Object {
    "error": Object {
      "message": "Expect C, got . ",
    },
    "inputPath": Array [],
    "resolverPath": Array [
      2,
    ],
  },
  Object {
    "error": Object {
      "message": "Expect D, got . ",
    },
    "inputPath": Array [],
    "resolverPath": Array [
      3,
    ],
  },
]
`;

exports[`union can validate 2`] = `
Array [
  Object {
    "error": Object {
      "message": "Expect A, got [object Object]. ",
    },
    "inputPath": Array [],
    "resolverPath": Array [
      0,
    ],
  },
  Object {
    "error": Object {
      "message": "Expect B, got [object Object]. ",
    },
    "inputPath": Array [],
    "resolverPath": Array [
      1,
    ],
  },
  Object {
    "error": Object {
      "message": "Expect C, got [object Object]. ",
    },
    "inputPath": Array [],
    "resolverPath": Array [
      2,
    ],
  },
  Object {
    "error": Object {
      "message": "Expect D, got [object Object]. ",
    },
    "inputPath": Array [],
    "resolverPath": Array [
      3,
    ],
  },
]
`;

exports[`union can validate 3`] = `
Array [
  Object {
    "error": Object {
      "message": "Expect M, got . ",
    },
    "inputPath": Array [
      "gender",
    ],
    "resolverPath": Array [
      "gender",
      0,
    ],
  },
  Object {
    "error": Object {
      "message": "Expect F, got . ",
    },
    "inputPath": Array [
      "gender",
    ],
    "resolverPath": Array [
      "gender",
      1,
    ],
  },
  Object {
    "error": Object {
      "message": "Expect Other, got . ",
    },
    "inputPath": Array [
      "gender",
    ],
    "resolverPath": Array [
      "gender",
      2,
    ],
  },
  Object {
    "error": Object {
      "message": "Expect a string, got number. ",
    },
    "inputPath": Array [
      "name",
    ],
    "resolverPath": Array [
      "name",
    ],
  },
]
`;
